File: artifacts/libzahl-1.0/translate_new_ablation/src/zstr_length.rs
use crate::*;
use crate::types::Z;
use crate::types::z_t;
use crate::zcmpmag;
use crate::zdiv;
use crate::zset;
use crate::zsetu;
use crate::zsignum;
use crate::zsqr;
use crate::zzero;
use lazy_static::lazy_static;

lazy_static! {
    pub static ref libzahl_tmp_str_div: std::sync::Mutex<z_t> =
        std::sync::Mutex::new([Z {
            sign: 0,
            used: 0,
            alloced: 0,
            chars: None,
        }]);
}

lazy_static! {
    pub static ref libzahl_tmp_str_mag: std::sync::Mutex<z_t> =
        std::sync::Mutex::new([Z {
            sign: 0,
            used: 0,
            alloced: 0,
            chars: None,
        }]);
}

lazy_static! {
    pub static ref libzahl_tmp_str_num: std::sync::Mutex<z_t> =
        std::sync::Mutex::new([Z {
            sign: 0,
            used: 0,
            alloced: 0,
            chars: None,
        }]);
}

pub fn zstr_length(a: &z_t, radix: u64) -> usize {
    let mut size_total = 1;
    
    {
        let mut libzahl_tmp_str_num_guard = libzahl_tmp_str_num.lock().unwrap();
        zset(&mut *libzahl_tmp_str_num_guard, a);
    }
    
    loop {
        let libzahl_tmp_str_num_guard = libzahl_tmp_str_num.lock().unwrap();
        if zzero(&*libzahl_tmp_str_num_guard) {
            break;
        }
        drop(libzahl_tmp_str_num_guard);
        
        {
            let mut libzahl_tmp_str_mag_guard = libzahl_tmp_str_mag.lock().unwrap();
            zsetu(&mut *libzahl_tmp_str_mag_guard, radix);
        }
        
        {
            let mut libzahl_tmp_str_div_guard = libzahl_tmp_str_div.lock().unwrap();
            let libzahl_tmp_str_mag_guard = libzahl_tmp_str_mag.lock().unwrap();
            zset(&mut *libzahl_tmp_str_div_guard, &*libzahl_tmp_str_mag_guard);
        }
        
        let mut size_temp = 1;
        
        loop {
            {
                let mut libzahl_tmp_str_div_guard = libzahl_tmp_str_div.lock().unwrap();
                let libzahl_tmp_str_mag_guard = libzahl_tmp_str_mag.lock().unwrap();
                zset(&mut *libzahl_tmp_str_div_guard, &*libzahl_tmp_str_mag_guard);
            }
            
            {
                let mut libzahl_tmp_str_mag_guard = libzahl_tmp_str_mag.lock().unwrap();
                let mut tmp_mag = [Z {
                    sign: 0,
                    used: 0,
                    alloced: 0,
                    chars: None,
                }];
                zset(&mut tmp_mag, &*libzahl_tmp_str_mag_guard);
                zsqr(&mut *libzahl_tmp_str_mag_guard, &mut tmp_mag);
            }
            
            size_temp <<= 1;
            
            let libzahl_tmp_str_mag_guard = libzahl_tmp_str_mag.lock().unwrap();
            let libzahl_tmp_str_num_guard = libzahl_tmp_str_num.lock().unwrap();
            if zcmpmag(&*libzahl_tmp_str_mag_guard, &*libzahl_tmp_str_num_guard) <= 0 {
                continue;
            } else {
                break;
            }
        }
        
        size_temp >>= 1;
        size_total += size_temp;
        
        {
            let mut libzahl_tmp_str_num_guard = libzahl_tmp_str_num.lock().unwrap();
            let libzahl_tmp_str_div_guard = libzahl_tmp_str_div.lock().unwrap();
            let mut tmp_num = [Z {
                sign: 0,
                used: 0,
                alloced: 0,
                chars: None,
            }];
            zset(&mut tmp_num, &*libzahl_tmp_str_num_guard);
            zdiv(&mut *libzahl_tmp_str_num_guard, &tmp_num, &*libzahl_tmp_str_div_guard);
        }
    }
    
    let sign = zsignum(a);
    size_total + if sign < 0 { 1 } else { 0 }
}

File: artifacts/libzahl-1.0/decompose/src/zstr_length.c
size_t zstr_length(z_t a, unsigned long long int radix)
{
  size_t size_total = 1;
  size_t size_temp;
  zset(libzahl_tmp_str_num, a);
  while (!zzero(libzahl_tmp_str_num))
  {
    zsetu(libzahl_tmp_str_mag, radix);
    zset(libzahl_tmp_str_div, libzahl_tmp_str_mag);
    size_temp = 1;
    while (zcmpmag(libzahl_tmp_str_mag, libzahl_tmp_str_num) <= 0)
    {
      zset(libzahl_tmp_str_div, libzahl_tmp_str_mag);
      zsqr(libzahl_tmp_str_mag, libzahl_tmp_str_mag);
      size_temp <<= 1;
    }

    size_temp >>= 1;
    size_total += size_temp;
    zdiv(libzahl_tmp_str_num, libzahl_tmp_str_num, libzahl_tmp_str_div);
  }

  return size_total + (zsignum(a) < 0);
}


/*** DEPENDENCIES:
extern z_t libzahl_tmp_str_mag
----------------------------
extern z_t libzahl_tmp_str_div
----------------------------
int zcmpmag(z_t a, z_t b)
{
  size_t i;
  size_t j;
  if (zzero(a))
  {
    return -(!zzero(b));
  }
  if (zzero(b))
  {
    return 1;
  }
  i = a->used - 1;
  j = b->used - 1;
  for (; i > j; i -= 1)
  {
    if (a->chars[i])
    {
      return +1;
    }
    a->used -= 1;
  }

  for (; j > i; j -= 1)
  {
    if (b->chars[j])
    {
      return -1;
    }
    b->used -= 1;
  }

  for (; i; i -= 1)
    if (a->chars[i] != b->chars[i])
  {
    return ((a->chars[i] > b->chars[i]) * 2) - 1;
  }

  return (a->chars[0] < b->chars[0]) ? (-1) : (a->chars[0] > b->chars[0]);
}


----------------------------
void zset(z_t a, z_t b)
{
  if (zzero(b))
  {
    a->sign = 0;
  }
  else
  {
    do
    {
      if (a->alloced < b->used)
      {
        libzahl_realloc(a, b->used);
      }
    }
    while (0);
    a->sign = b->sign;
    a->used = b->used;
    memcpy(a->chars, b->chars, b->used * (sizeof(zahl_char_t)));
  }
}


----------------------------
inline static int zzero(z_t a)
{
  return !a->sign;
}


----------------------------
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1]
----------------------------
void zsetu(z_t a, unsigned long long int b)
{
  if (!b)
  {
    a->sign = 0;
    return;
  }
  do
  {
    if (a->alloced < ((((sizeof(b)) + (sizeof(*a->chars))) - 1) / (sizeof(*a->chars))))
    {
      libzahl_realloc(a, (((sizeof(b)) + (sizeof(*a->chars))) - 1) / (sizeof(*a->chars)));
    }
  }
  while (0);
  a->sign = 1;
  a->used = 0;
  while (b)
  {
    a->chars[a->used++] = (zahl_char_t) b;
    b >>= 32;
  }

}


----------------------------
void zsqr(z_t a, z_t b)
{
  size_t m2;
  z_t z0;
  z_t z1;
  z_t z2;
  z_t high;
  z_t low;
  int sign;
  if (zzero(b))
  {
    a->sign = 0;
    return;
  }
  m2 = zbits(b);
  if (m2 <= (32 / 2))
  {
    do
    {
      if (a->alloced < 1)
      {
        libzahl_realloc(a, 1);
      }
    }
    while (0);
    a->used = 1;
    a->chars[0] = b->chars[0] * b->chars[0];
    a->sign = 1;
    return;
  }
  sign = zsignum(b);
  b->sign = 1;
  m2 >>= 1;
  zinit(z0);
  zinit(z1);
  zinit(z2);
  zinit(high);
  zinit(low);
  zsplit(high, low, b, m2);
  zsqr(z0, low);
  zsqr(z2, high);
  zmul(z1, low, high);
  zlsh(z1, z1, m2 + 1);
  m2 <<= 1;
  zlsh(z2, z2, m2);
  zadd(a, z2, z1);
  zadd(a, a, z0);
  zfree(z0);
  zfree(z1);
  zfree(z2);
  zfree(high);
  zfree(low);
  b->sign = sign;
  a->sign = 1;
}


----------------------------
void zdiv(z_t a, z_t b, z_t c)
{
  zdivmod(a, libzahl_tmp_div, b, c);
}


----------------------------
extern z_t libzahl_tmp_str_num
----------------------------
inline static int zsignum(z_t a)
{
  return a->sign;
}


----------------------------
***/


